File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/gamespots/TicTacToeGame.java
package systems.zlink.samples.tictactoe.server.play.gamespots;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import systems.zlink.contracts.messaging.Message;
import systems.zlink.framework.CancellationToken;
import systems.zlink.framework.actors.ZLinkActor;
import systems.zlink.framework.spots.ZLinkSpot;
import systems.zlink.framework.spots.ZLinkSpotActorJoinResponse;
import systems.zlink.framework.spots.ZLinkSpotContext;
import systems.zlink.framework.spots.ZLinkSpotCreateResponse;
import systems.zlink.framework.spots.ZLinkTimer;
import systems.zlink.framework.spots.ZLinkTimerOptions;
import systems.zlink.samples.tictactoe.server.play.actors.PlayActor;
import systems.zlink.samples.tictactoe.server.play.gamespots.handlers.TicTacToeGameCreatedHandler;
import systems.zlink.samples.tictactoe.server.play.gamespots.handlers.TicTacToeGameTimerHandler;
import systems.zlink.samples.tictactoe.shared.contracts.GameState;
import systems.zlink.samples.tictactoe.shared.contracts.GameStateNotify;
import systems.zlink.samples.tictactoe.shared.contracts.PlaceMarkRes;
import systems.zlink.samples.tictactoe.shared.contracts.PlayerJoinedNotify;
import systems.zlink.samples.tictactoe.shared.contracts.TicTacToeGameJoinReq;
import systems.zlink.samples.tictactoe.shared.contracts.TicTacToeGameJoinRes;

public final class TicTacToeGame implements ZLinkSpot {
    private static final Duration GAME_TICK_PERIOD = Duration.ofSeconds(1);
    private static final Duration TURN_TIMEOUT = Duration.ofSeconds(15);

    private final ZLinkSpotContext context;
    private final String gameId;
    private final char[] board = ".........".toCharArray();
    private final List<PlayerSlot> players = new ArrayList<>();
    private String status = "WaitingForPlayers";
    private String nextTurn = "X";
    private String winner = "";
    private String lastMoveActorId = "";
    private int lastMoveCell = -1;
    private Instant turnDeadline;
    private ZLinkTimer gameTick;
    private boolean created;
    private final TicTacToeGameCreatedHandler createdHandler;
    private final ObjectMapper json;

    public TicTacToeGame(
        ZLinkSpotContext context,
        TicTacToeGameCreatedHandler createdHandler,
        ObjectMapper json) {
        this.context = context;
        this.gameId = context.spotRid().toHex();
        this.createdHandler = createdHandler;
        this.json = json;
    }

    public String gameId() {
        return gameId;
    }

    @Override
    public ZLinkSpotContext context() {
        return context;
    }

    @Override
    public CompletionStage<ZLinkSpotCreateResponse> onCreateAsync(Message request) {
        return createdHandler.handleAsync(this, request);
    }

    @Override
    public CompletionStage<ZLinkSpotActorJoinResponse> onActorJoinAsync(
        ZLinkActor actor,
        Message request,
        CancellationToken cancellationToken) {
        if (!(actor instanceof PlayActor player)) {
            return CompletableFuture.failedFuture(
                new IllegalArgumentException("tic-tac-toe game only accepts PlayActor."));
        }
        TicTacToeGameJoinReq joinRequest = decode(request, TicTacToeGameJoinReq.class);
        return join(player, joinRequest.gameId())
            .thenApply(reply -> ZLinkSpotActorJoinResponse.accept(encode(reply)));
    }

    @Override
    public CompletionStage<Void> onPostActorJoinedAsync(
        ZLinkActor actor,
        CancellationToken cancellationToken) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onActorLeftAsync(
        ZLinkActor actor,
        CancellationToken cancellationToken) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onInitializeAsync() {
        return context.addTimer(
                "game-tick",
                GAME_TICK_PERIOD,
                TicTacToeGameTimerHandler.class,
                new ZLinkTimerOptions())
            .thenAccept(timer -> gameTick = timer);
    }

    @Override
    public CompletionStage<Void> onClosingAsync() {
        return gameTick == null
            ? CompletableFuture.completedFuture(null)
            : gameTick.cancelAsync();
    }

    public void markCreated(Message request) {
        if (!request.isEmpty()) {
            throw new IllegalArgumentException("tic-tac-toe game creation does not accept payload");
        }
        created = true;
    }

    public CompletionStage<TicTacToeGameJoinRes> join(PlayActor actor, String gameId) {
        ensureCreated();
        PlayerSlot slot = players.stream()
            .filter(player -> player.actor().actorId().equals(actor.actorId()))
            .findFirst()
            .orElse(null);
        boolean isNewPlayer = slot == null;
        if (slot == null) {
            if (players.size() >= 2) {
                throw new IllegalStateException("tic-tac-toe game already has two players");
            }
            String mark = players.size() == 0 ? "X" : "O";
            slot = new PlayerSlot(actor, mark);
            players.add(slot);
        } else {
            slot.actor(actor);
        }
        if (players.size() == 2 && status.equals("WaitingForPlayers")) {
            status = "InProgress";
            resetTurnDeadline();
        }
        actor.joinGame(gameId);
        GameState state = snapshot();
        CompletionStage<Void> notify = isNewPlayer
            ? notifyPlayerJoined(actor, slot, state)
            : CompletableFuture.completedFuture(null);
        return notify
            .thenCompose(ignored -> broadcast(state, actor.actorId()))
            .thenApply(ignored -> new TicTacToeGameJoinRes(state));
    }

    public CompletionStage<PlaceMarkRes> placeMark(PlayActor actor, int cell) {
        ensureCreated();
        PlayerSlot slot = players.stream()
            .filter(player -> player.actor().actorId().equals(actor.actorId()))
            .findFirst()
            .orElseThrow(() -> new IllegalStateException("player has not joined"));
        if (!status.equals("InProgress")) {
            throw new IllegalStateException("game is not in progress");
        }
        if (!slot.mark().equals(nextTurn)) {
            throw new IllegalStateException("unexpected turn");
        }
        if ((uint(cell)) >= board.length || board[cell] != '.') {
            throw new IllegalArgumentException("invalid cell");
        }

        board[cell] = slot.mark().charAt(0);
        lastMoveActorId = actor.actorId();
        lastMoveCell = cell;
        advance(slot);
        GameState state = snapshot();
        return broadcast(state, actor.actorId())
            .thenApply(ignored -> new PlaceMarkRes(state));
    }

    public String winner() {
        return winner;
    }

    public boolean hasPlayer(String actorId) {
        return players.stream().anyMatch(player -> player.actor().actorId().equals(actorId));
    }

    public GameState snapshot() {
        ensureCreated();
        return new GameState(
            gameId,
            new String(board),
            status,
            winner.isEmpty() ? null : winner,
            nextTurn,
            actorIdForMark("X"),
            actorIdForMark("O"),
            lastMoveActorId.isEmpty() ? null : lastMoveActorId,
            lastMoveCell < 0 ? null : lastMoveCell);
    }

    private void advance(PlayerSlot slot) {
        if (hasWon(slot.mark().charAt(0))) {
            status = "Won";
            winner = slot.actor().actorId();
            nextTurn = "";
            turnDeadline = null;
        } else if (new String(board).indexOf('.') < 0) {
            status = "Draw";
            winner = "";
            nextTurn = "";
            turnDeadline = null;
        } else {
            nextTurn = slot.mark().equals("X") ? "O" : "X";
            resetTurnDeadline();
        }
    }

    public CompletionStage<Void> tickAsync() {
        ensureCreated();
        if (!status.equals("InProgress")
            || turnDeadline == null
            || Instant.now().isBefore(turnDeadline)) {
            return CompletableFuture.completedFuture(null);
        }

        PlayerSlot timedOut = players.stream()
            .filter(player -> player.mark().equals(nextTurn))
            .findFirst()
            .orElse(null);
        PlayerSlot winningSlot = players.stream()
            .filter(player -> !player.mark().equals(nextTurn))
            .findFirst()
            .orElse(null);

        status = "TurnTimedOut";
        winner = winningSlot == null ? "" : winningSlot.actor().actorId();
        nextTurn = "";
        lastMoveActorId = timedOut == null ? "" : timedOut.actor().actorId();
        lastMoveCell = -1;
        turnDeadline = null;

        return broadcast(snapshot(), null);
    }

    private void resetTurnDeadline() {
        turnDeadline = Instant.now().plus(TURN_TIMEOUT);
    }

    private void ensureCreated() {
        if (!created) {
            throw new IllegalStateException("tic-tac-toe game has not completed creation");
        }
    }

    private <T> T decode(Message request, Class<T> type) {
        try {
            return json.readValue(request.toByteArray(), type);
        } catch (java.io.IOException ex) {
            throw new IllegalArgumentException("failed to decode " + type.getSimpleName(), ex);
        }
    }

    private Message encode(Object reply) {
        try {
            return Message.from(json.writeValueAsBytes(reply));
        } catch (JsonProcessingException ex) {
            throw new IllegalArgumentException("failed to encode tic-tac-toe join reply", ex);
        }
    }

    private String actorIdForMark(String mark) {
        return players.stream()
            .filter(player -> player.mark().equals(mark))
            .map(player -> player.actor().actorId())
            .findFirst()
            .orElse(null);
    }

    private CompletionStage<Void> broadcast(GameState state, String excludedActorId) {
        List<CompletionStage<Void>> sends = players.stream()
            .map(PlayerSlot::actor)
            .filter(actor -> excludedActorId == null || !actor.actorId().equals(excludedActorId))
            .map(actor -> actor.context().boundSession()
                .send(new GameStateNotify(state))
                .submitAsync())
            .toList();
        return allOf(sends);
    }

    private CompletionStage<Void> notifyPlayerJoined(
        PlayActor joinedActor,
        PlayerSlot joinedSlot,
        GameState state) {
        PlayerJoinedNotify message = new PlayerJoinedNotify(
            state.gameId(),
            joinedActor.actorId(),
            joinedSlot.mark(),
            state);
        List<CompletionStage<Void>> sends = players.stream()
            .map(PlayerSlot::actor)
            .filter(actor -> !actor.actorId().equals(joinedActor.actorId()))
            .map(actor -> actor.context().boundSession()
                .send(message)
                .submitAsync())
            .toList();
        return allOf(sends);
    }

    private static CompletionStage<Void> allOf(List<CompletionStage<Void>> stages) {
        CompletionStage<Void> combined = CompletableFuture.completedFuture(null);
        for (CompletionStage<Void> stage : stages) {
            combined = combined.thenCompose(ignored -> stage);
        }
        return combined;
    }

    private boolean hasWon(char mark) {
        int[][] lines = {
            {0, 1, 2}, {3, 4, 5}, {6, 7, 8},
            {0, 3, 6}, {1, 4, 7}, {2, 5, 8},
            {0, 4, 8}, {2, 4, 6}
        };
        for (int[] line : lines) {
            if (board[line[0]] == mark && board[line[1]] == mark && board[line[2]] == mark) {
                return true;
            }
        }
        return false;
    }

    private static int uint(int value) {
        return value < 0 ? Integer.MAX_VALUE : value;
    }

    private static final class PlayerSlot {
        private PlayActor actor;
        private final String mark;

        private PlayerSlot(PlayActor actor, String mark) {
            this.actor = actor;
            this.mark = mark;
        }

        private PlayActor actor() {
            return actor;
        }

        private void actor(PlayActor actor) {
            this.actor = actor;
        }

        private String mark() {
            return mark;
        }
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/infrastructure/zlink/spots/TicTacToeGame.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots

import com.fasterxml.jackson.databind.ObjectMapper
import java.time.Duration
import java.time.Instant
import kotlinx.coroutines.future.await
import systems.zlink.contracts.messaging.Message
import systems.zlink.framework.CancellationToken
import systems.zlink.framework.ZLinkAwait
import systems.zlink.framework.kotlin.ZLinkSuspendingSpot
import systems.zlink.framework.spots.ZLinkSpotActorJoinResponse
import systems.zlink.framework.spots.ZLinkSpotContext
import systems.zlink.framework.spots.ZLinkSpotCreateResponse
import systems.zlink.framework.spots.ZLinkTimer
import systems.zlink.framework.spots.ZLinkTimerOptions
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.actors.PlayActor
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.handlers.TicTacToeGameCreatedHandler
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.handlers.TicTacToeGameTimerHandler
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.GameState
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.GameStateNotify
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlaceMarkRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerInfo
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerJoinedNotify
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerWinMilestoneEvent
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.TicTacToeGameJoinReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.TicTacToeGameJoinRes

class TicTacToeGame(
    private val context: ZLinkSpotContext,
    private val createdHandler: TicTacToeGameCreatedHandler,
    private val json: ObjectMapper,) : ZLinkSuspendingSpot<PlayActor>() {
    private val gameTickPeriod: Duration = Duration.ofSeconds(1)
    private val turnTimeout: Duration = Duration.ofSeconds(15)
    val roomId: String = context.spotRid().toString()
    private val board = ".........".toCharArray()
    private val players = mutableListOf<PlayerSlot>()
    private var status = "WaitingForPlayers"
    private var nextTurn = "X"
    private var winnerValue: String? = null
    private var lastMoveActorId: String? = null
    private var lastMoveCell: Int? = null
    private var turnDeadline: Instant? = null
    private var gameTick: ZLinkTimer? = null
    private var created = false
    private var cleanupStarted = false

    override fun context(): ZLinkSpotContext = context

    override suspend fun onCreateSuspending(request: Message): ZLinkSpotCreateResponse {
        createdHandler.handle(this, request)
        return ZLinkSpotCreateResponse.accept()
    }

    override suspend fun onActorJoinSuspending(
        actor: PlayActor,
        request: Message,
        cancellationToken: CancellationToken,
    ): ZLinkSpotActorJoinResponse {
        val joinRequest = json.readValue(request.toByteArray(), TicTacToeGameJoinReq::class.java)
        require(joinRequest.player.actorId == actor.actorId) {
            "join request actor id does not match bound actor"
        }
        val reply = join(actor, joinRequest.roomId, joinRequest.player)
        return ZLinkSpotActorJoinResponse.accept(Message.from(json.writeValueAsBytes(reply)))
    }

    override fun onJoinedActor(
        actor: PlayActor,
        cancellationToken: CancellationToken,
    ) {
    }

    override fun onLeaveActor(
        actor: PlayActor,
        cancellationToken: CancellationToken,
    ) {
        players.removeIf { it.actor.actorId == actor.actorId }
    }

    override fun onDisconnectActor(
        actor: PlayActor,
        cancellationToken: CancellationToken,
    ) {
        actor.markDisconnected()
    }

    override suspend fun onInitializeSuspending() {
        gameTick = context.addTimer(
            "game-tick",
            gameTickPeriod,
            TicTacToeGameTimerHandler::class.java,
            ZLinkTimerOptions(),
        ).await()
    }

    override suspend fun onClosingSuspending() {
        gameTick?.cancelAsync()?.await()
    }

    fun markCreated(request: Message) {
        require(request.isEmpty()) { "tic-tac-toe game creation does not accept payload parts" }
        created = true
    }

    suspend fun join(actor: PlayActor, roomId: String, player: PlayerInfo): TicTacToeGameJoinRes {
        ensureCreated()
        check(roomId == this.roomId) { "join request room id does not match game room" }
        check(player.level >= SampleNames.RequiredLevel) { "player level does not satisfy room requirement" }
        actor.applyPlayer(player)
        var slot = players.firstOrNull { it.actor.actorId == actor.actorId }
        val isNewPlayer = slot == null
        if (slot == null) {
            check(players.size < 2) { "tic-tac-toe game already has two players" }
            val mark = if (players.size == 0) "X" else "O"
            slot = PlayerSlot(actor, mark)
            players += slot
        } else {
            slot.actor = actor
        }
        if (players.size == 2 && status == "WaitingForPlayers") {
            status = "InProgress"
            resetTurnDeadline()
        }
        actor.joinGame(roomId)
        val state = snapshot()
        if (isNewPlayer) {
            notifyPlayerJoined(actor, slot, state)
        }
        broadcast(state, actor.actorId)
        return TicTacToeGameJoinRes(state)
    }

    suspend fun placeMark(actor: PlayActor, cell: Int): PlaceMarkRes {
        ensureCreated()
        val slot = players.firstOrNull { it.actor.actorId == actor.actorId }
            ?: throw IllegalStateException("player has not joined")
        check(status == "InProgress") { "game is not in progress" }
        check(slot.mark == nextTurn) { "unexpected turn" }
        require(cell in board.indices && board[cell] == '.') { "invalid cell" }

        val before = snapshot()
        board[cell] = slot.mark[0]
        lastMoveActorId = actor.actorId
        lastMoveCell = cell
        advance(slot)
        val state = snapshot()
        broadcast(state, actor.actorId)
        publishWinMilestone(actor, before, state)
        return PlaceMarkRes(state)
    }

    fun hasPlayer(actorId: String): Boolean =
        players.any { it.actor.actorId == actorId }

    private fun snapshot(): GameState {
        ensureCreated()
        return GameState(
            roomId = roomId,
            board = String(board),
            status = status,
            winner = winnerValue,
            nextTurn = nextTurn,
            xActorId = players.firstOrNull { it.mark == "X" }?.actor?.actorId,
            oActorId = players.firstOrNull { it.mark == "O" }?.actor?.actorId,
            lastMoveActorId = lastMoveActorId,
            lastMoveCell = lastMoveCell,
        )
    }

    private fun advance(slot: PlayerSlot) {
        if (hasWon(slot.mark[0])) {
            status = "Won"
            winnerValue = slot.actor.actorId
            nextTurn = ""
            turnDeadline = null
        } else if ('.' !in board) {
            status = "Draw"
            winnerValue = null
            nextTurn = ""
            turnDeadline = null
        } else {
            nextTurn = if (slot.mark == "X") "O" else "X"
            resetTurnDeadline()
        }
    }

    suspend fun tick() {
        ensureCreated()
        val deadline = turnDeadline
        if (status != "InProgress" || deadline == null || Instant.now().isBefore(deadline)) {
            leaveFinishedActors(snapshot())
            return
        }

        val timedOut = players.firstOrNull { it.mark == nextTurn }
        val winner = players.firstOrNull { it.mark != nextTurn }

        status = "TurnTimedOut"
        winnerValue = winner?.actor?.actorId
        nextTurn = ""
        lastMoveActorId = timedOut?.actor?.actorId
        lastMoveCell = null
        turnDeadline = null

        val state = snapshot()
        broadcast(state, null)
        leaveFinishedActors(state)
    }

    private suspend fun leaveFinishedActors(state: GameState) {
        if (cleanupStarted || !isTerminal(state)) {
            return
        }
        cleanupStarted = true
        for (slot in players.toList()) {
            slot.actor.markForDestroyAfterRoomLeave()
            context.leaveActor(slot.actor).await()
        }
    }

    private fun resetTurnDeadline() {
        turnDeadline = Instant.now().plus(turnTimeout)
    }

    private fun ensureCreated() {
        check(created) { "tic-tac-toe game has not completed creation" }
    }

    private fun hasWon(mark: Char): Boolean {
        val lines = arrayOf(
            intArrayOf(0, 1, 2),
            intArrayOf(3, 4, 5),
            intArrayOf(6, 7, 8),
            intArrayOf(0, 3, 6),
            intArrayOf(1, 4, 7),
            intArrayOf(2, 5, 8),
            intArrayOf(0, 4, 8),
            intArrayOf(2, 4, 6),
        )
        return lines.any { (a, b, c) -> board[a] == mark && board[b] == mark && board[c] == mark }
    }

    private suspend fun broadcast(state: GameState, excludedActorId: String?) {
        players
            .asSequence()
            .map { it.actor }
            .filter { excludedActorId == null || it.actorId != excludedActorId }
            .forEach { actor ->
                actor.context().boundSession()
                    .send(GameStateNotify(state))
                    .submit()
                    .await()
            }
    }

    private suspend fun notifyPlayerJoined(
        joinedActor: PlayActor,
        joinedSlot: PlayerSlot,
        state: GameState,
    ) {
        val player = joinedActor.requirePlayer()
        val message = PlayerJoinedNotify(
            roomId = state.roomId,
            actorId = joinedActor.actorId,
            displayName = player.displayName,
            level = player.level,
            mark = joinedSlot.mark,
            state = state,
        )
        players
            .asSequence()
            .map { it.actor }
            .filter { it.actorId != joinedActor.actorId }
            .forEach { actor ->
                actor.context().boundSession()
                    .send(message)
                    .submit()
                    .await()
            }
    }

    private data class PlayerSlot(var actor: PlayActor, val mark: String)

    private fun isTerminal(state: GameState): Boolean =
        state.status == "TurnTimedOut"

    fun leaveGame(actor: PlayActor, roomId: String) {
        check(this.roomId == roomId) { "leave request room id does not match game room" }
        actor.markForDestroyAfterRoomLeave()
        ZLinkAwait.await(context.leaveActor(actor))
        println("actor: LeaveGameReq completed. actor=${actor.actorId}")
    }

    private suspend fun publishWinMilestone(
        actor: PlayActor,
        before: GameState,
        after: GameState,
    ) {
        if (after.status != "Won" || before.status == "Won" || after.winner != actor.actorId) {
            return
        }
        val player = actor.requirePlayer()
        val wins = actor.incrementWins()
        if (player.wins < 99 || wins != 100) {
            return
        }
        context.outbound()
            .publish(
                SampleNames.PlayerMilestoneTopic,
                PlayerWinMilestoneEvent(
                    roomId = after.roomId,
                    actorId = actor.actorId,
                    displayName = player.displayName,
                    wins = wins,
                ),
            )
            .await()
    }
}
